src/main/java/duke/ui/Ui.java

```diff
@@ -0,0 +1,147 @@
+package duke.ui;
+
+import java.util.ArrayList;
+import java.util.Scanner;
+
+/**
+ * The `Ui` class is responsible for the interactions with the user.
+ */
+public class Ui {
+
+    private static final String GREETING = "Hello! I'm JED, your personal chat-bot!\nWhat can I do for you?";
+    private static final String GOODBYE = "Bye. Hope to see you again soon!";
+    private static final String EMPTY_LIST = "Your list is currently empty.";
+    private static final String ERROR_CANNOT_LOAD_DATA =
+            "I am unable to read your duke.txt data file. Starting with a blank task list...";
+    private static final String NO_MATCHING_TASKS = "Your keyword search returned no matching tasks.";
+    private Scanner sc;
+
+    /**
+     * Constructs a `Ui` object.
+     */
+    public Ui() {
+        this.sc = new Scanner(System.in);
+    }
+
+    /**
+     * Retrieves the next line of input from the user and returns it.
+     *
+     * @return The next line of input.
+     */
+    public String getInput() {
+        return this.sc.nextLine();
+    }
+
+    /**
+     * Prints out the specified String, wrapped in a horizontal line break.
+     *
+     * @param str The string to be printed out.
+     * @return The string to be printed out.
+     */
+    public String talk(String str) {
+        String line = "_".repeat(50);
+        System.out.println(line);
+        System.out.println(str);
+        System.out.println(line);
+        return str;
+    }
+
+    /**
+     * Prints out the greeting message.
+     */
+    public String greet() {
+        return talk(GREETING);
+    }
+
+    /**
+     * Prints out the goodbye message.
+     */
+    public String bye() {
+        return talk(GOODBYE);
+        // sc.close();
+    }
+
+    /**
+     * Prints out the loading error message, used when Duke is unable to read the
+     * input data file.
+     */
+    public String showLoadingError() {
+        return talk(ERROR_CANNOT_LOAD_DATA);
+    }
+
+    /**
+     * Prints out the specified ArrayList of items in the standard format, where
+     * items are 1-indexed.
+     *
+     * @param items The ArrayList of items to be printed.
+     */
+    public String list(ArrayList<?> items) {
+        int count = items.size();
+        if (count == 0) {
+            return talk(EMPTY_LIST);
+        } else {
+            String list = "";
+            for (int i = 0; i < count; i++) {
+                list += "  " + (i + 1) + ". " + items.get(i) + "\n";
+            }
+            return talk(list);
+        }
+
+    }
+
+    /**
+     * Prints out the specified ArrayList of searched items in the standard format,
+     * where items are 1-indexed.
+     *
+     * @param items The ArrayList of searched items to be printed.
+     */
+    public String listSearch(ArrayList<?> items) {
```

Comment:
Why does the `items` parameter here need to be of type `ArrayList<?>`? Specifically confused about the need for the wildcard generic in the ArrayList.